### PR TITLE
Defend empty commit message and Add test case.

### PR DESCRIPTION
**Pull request by dlworldpeace**
As I was testing RepoSense on one of my module project repo at this link https://github.com/dlworldpeace/SleaseWeb/commits/master
it encounters ArrayIndexOutOfBoundException when empty commit message is used. 
Thus, this PR aims to fix it and add an empty commit message test case.

### CODE REVIEW

**Comment by eugenepeh** [comment]
Hi @dlworldpeace,

Before you start working on a PR, can you please open an issue and provide us with the full information?

Thanks!

File: src/main/java/reposense/commits/CommitInfoAnalyzer.java [hunk 1]
```diff
@@ -67,7 +67,13 @@ public static CommitResult analyzeCommit(CommitInfo commitInfo, Map<String, Auth
             logger.log(Level.WARNING, "Unable to parse the date from git log result for commit.", pe);
         }
 
-        String message = elements[MESSAGE_INDEX];
+        String message = "";
+        try {
```
Comment (ongspxm) [inline src/main/java/reposense/commits/CommitInfoAnalyzer.java]: will using `elements.length` to do a check instead of catching the exception be a better solution?

Comment (yamidark) [reply src/main/java/reposense/commits/CommitInfoAnalyzer.java]: Yes, we typically only use `try/catch` of exceptions for situations that are 'exceptional'/out of our control.

In this case, an empty commit message isn't exactly 'exceptional', but just 1 special/corner case. As such, it is preferred to use a `if` statement to check for the array's length, and setting the message to be empty if it is out of range. We also don't need a logging message for this case as well.

Comment (dlworldpeace) [reply src/main/java/reposense/commits/CommitInfoAnalyzer.java]: Thanks  for pointing out. Have updated accordingly.

**Comment by yong24s** [comment]
@dlworldpeace 

Please use `https://github.com/reposense/testrepo-EmptyCommitMessage` in your test.

**Comment by eugenepeh** [comment]
@yong24s for unit test, all should be in testrepo-Alpha for standardization. 
can use a new branch for it.

**Comment by dlworldpeace** [comment]
> @dlworldpeace
> 
> Please use `https://github.com/reposense/testrepo-Alpha` with `empty-commit-message` branch in your test.

Thanks for pointing it out. However, may I know why I am only seeing testrepo-Delta repo as shown:
image

**Comment by yong24s** [comment]
@dlworldpeace 

Not sure why, we use git cli but it can be accessed by `https://github.com/reposense/testrepo-Alpha/tree/empty-commit-message`.

File: src/main/java/reposense/commits/CommitInfoAnalyzer.java [hunk 2]
```diff
@@ -67,7 +68,11 @@ public static CommitResult analyzeCommit(CommitInfo commitInfo, Map<String, Auth
             logger.log(Level.WARNING, "Unable to parse the date from git log result for commit.", pe);
         }
 
-        String message = elements[MESSAGE_INDEX];
+        String message = "";
+        if (elements.length > ELEMENTS_LENGTH_IF_EMPTY_MESSAGE) {
```
Comment (yong24s) [inline src/main/java/reposense/commits/CommitInfoAnalyzer.java]: It seems that you intend to use const `ELEMENTS_LENGTH_IF_EMPTY_MESSAGE` to simplify expression.

Try using a boolean instead.

```
boolean containsMessage =  elements.length > MESSAGE_INDEX;
String message = (containsMessage) ? elements[MESSAGE_INDEX] : "";
```

**Comment by dlworldpeace** [comment]
@reposense/devs ready for next round of review.

**Comment by dlworldpeace** [comment]
> @dlworldpeace
> 
> Please use `https://github.com/reposense/testrepo-Alpha` with `empty-commit-message` branch in your test.

thanks! already created PR for a test with empty commit message.

**Comment by dlworldpeace** [comment]
> @dlworldpeace
> 
> @yong24s Have already added a commit without commit message in the `empty-commit-message` branch of testrepo-Alpha. Do use that and add the appropriate test cases here.

test case added, ready for review.

File: src/main/java/reposense/model/Format.java [hunk 3]
```diff
@@ -9,7 +9,8 @@
  */
 public class Format {
     public static final List<String> DEFAULT_FORMAT_STRINGS = Arrays.asList(
-            "adoc", "cs", "css", "fxml", "gradle", "html", "java", "js", "json", "jsp", "md", "py", "tag", "xml");
+            "adoc", "cs", "css", "fxml", "gradle", "html", "java", "js", "json",
+            "jsp", "md", "py", "tag", "xml", "txt");
```
Comment (eugenepeh) [inline src/main/java/reposense/model/Format.java]: hmm, interesting addition, seems like we missed out `txt` file. 
However, this PR does not warrant this fix; each PR should only target a single issue (unless they coincides, not in this sense though) to keep it history of changes clean.

Hence, please remove this change and open another issue for it.

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java [hunk 4]
```diff
@@ -84,4 +85,17 @@ public void analyzeCommits_allAuthorMultipleCommitIgnoredNoDateRange_success() {
         Assert.assertEquals(commitResultsShort, commitResultsFull);
         Assert.assertEquals(commitInfos.size() - 2, commitResultsFull.size());
     }
+
+    @Test
+    public void analyzeCommits_yongAuthorNoCommitIgnoredNoDateRangeEmptyMessage_success() {
```
Comment (eugenepeh) [inline src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: method title too long, can just leave it as `analyzeCommits_noCommitMessage_success`

File: src/main/java/reposense/model/Format.java (hunk 3 above)
Comment (yong24s) [reply src/main/java/reposense/model/Format.java]: Opps 😱 

Try this:

```
config.setFormats(Format.convertStringsToFormats(Arrays.asList("txt")));
```

**Comment by dlworldpeace** [comment]
> Great work so far, just a few changes

Updated accordingly, ready for review.

File: src/main/java/reposense/commits/CommitInfoAnalyzer.java [hunk 5]
```diff
@@ -68,7 +68,8 @@ public static CommitResult analyzeCommit(CommitInfo commitInfo, Map<String, Auth
             logger.log(Level.WARNING, "Unable to parse the date from git log result for commit.", pe);
         }
 
-        String message = elements[MESSAGE_INDEX];
+        boolean containsMessage = elements.length > MESSAGE_INDEX;
+        String message = (containsMessage) ? elements[MESSAGE_INDEX] : "";
```
Comment (yamidark) [inline src/main/java/reposense/commits/CommitInfoAnalyzer.java]: I think we can put this 2 into a single line

Comment (dlworldpeace) [reply src/main/java/reposense/commits/CommitInfoAnalyzer.java]: Do you mean 'String message = (elements.length > MESSAGE_INDEX) ? elements[MESSAGE_INDEX] : "";'?

Comment (yamidark) [reply src/main/java/reposense/commits/CommitInfoAnalyzer.java]: Yes

**Comment by eugenepeh** [comment]
Thanks @dlworldpeace ,

Your PR is ready to be merged. All now that is left to do is to construct your merge commit message as specified in our commit organization.

Please follow the following flow as specified in our [commit organization](https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message) as specified in the [contribution guidelines](https://github.com/reposense/RepoSense/blob/master/docs/Process.md#how-to-contribute-to-the-reposense-repository):
```
{current situation} -- use present tense

{why it needs to change}

{what is being done about it} -- use imperative mood

{why it is done that way}

{any other relevant info}
```

You may also refer to [past commits](https://github.com/reposense/RepoSense/commits/master) to help you in constructing the merge commit message.

**Comment by dlworldpeace** [comment]
Ready for last review before merge.

**Comment by eugenepeh** [comment]
Nits to the proposed message:

```
We analyzes the raw commit history to obtain information, such as
author, hash, date, message, insertion, deletion etc, about the
commit. The raw data undergoes splitting in order to obtain internal
conceptual representation. However, in the event of encountering a
commit with empty messages, the last index of split elements would be
missing and cause an ArrayIndexOutOfBoundException to be thrown. 

Although, by default, git does not allow empty commit message but this
behavior can be override by this command `git commit
--allow-empty-message`. Also, lightweight IDEs like Visual Studio Code
do allow users to commit without any message by accident.

To prevent breakage due to these corner cases, let's perform an index
check on the elements to avoid RepoSense crash due to the uncaught
exception.

Additional info:
A corner test case on this is included. It is made on the
empty-commit-message branch of testrepo-Alpha.

When committing changes on the Visual Studio Code, if only white spaces
were entered in commit message box, the IDE will commit the changes
without a message.
```

**Comment by eugenepeh** [comment]
@yamidark can you help to review the proposed message?

**Comment by yamidark** [comment]
>@yamidark can you help to review the proposed message?

LGTM 👍

**Comment by yamidark** [comment]
@dlworldpeace Seems like there are some conflicts with the latest master, do resolve these and update your PR branch.
